### ISSUE TITLE
[py2py3] Remove/replace dependency on python-cjson library

### ISSUE THREAD

**Issue by mapellidario** [dmwm/WMCore#10075]
**Impact of the new feature**

WMCore in general, and perhaps other services depending on it.

**Is your feature request related to a problem? Please describe.**

Some packages built from dmwm/WMCore depend on `python-cjson` ( [pypi](https://pypi.org/project/python-cjson/), [specfile](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/py2-cjson.spec) ), which only supports python2. 

List of dmwm/WMCore packages depending on `python-cjson`

* [WMagent](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/wmagent.spec), imported by [dbs3-client](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/dbs3-client.specc).
* [ReqMgr2](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/reqmgr2.spec), imported by [dbs3-client](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/dbs3-client.specc).
* [Workqueue](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/workqueue.spec), imported by [dbs3-client](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/dbs3-client.specc).
* [Reqmon](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/reqmon.spec), imported by [dbs3-client](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/dbs3-client.specc).

**Describe the solution you'd like**

The issue about the migration of `python-cjson` to python3 is still [open](https://github.com/AGProjects/python-cjson/issues/6) after 3 years. It seems unlikely that this changes soon.

We could use standard json library, but that would require a bit performance profiling of WMCore code.

**Comment by vkuznet** [comment]
The usage of cjson python module was chosen to increase performance of DBS code. Removing or replacing this may have significant impact on DBS code performance. I suggest that this issue should be carefully discussed with @yuyiguo . In a past we did several studies of using different json implementation for python and choice of cjson module was based on them.

**Comment by klannon** [comment]
@vkuznet: Thanks for the reminder and for mentioning @yuyiguo.  For this case, I think the first step is to explore alternatives.  From just about 10 minutes of Googling it seems clear to me that [cjson](https://pypistats.org/packages/python-cjson) is not really well supported while there are other options that are more widely used and more actively supported (including full Py3 support already today).  Two that I found include [orjson](https://pypistats.org/packages/orjson) and [rapidjson](https://pypistats.org/packages/python-rapidjson).  I don't have any performance benchmarking on these, so the first step would be to compare these or some other well-supported Py3 json package to our current cjson choice to see how they compare.

**Comment by vkuznet** [comment]
Kevin, it was a very long story and we did lots of benchmark. Here is an email from 2009 which outlines the issue:
```
I did some testing. Here are the results:

cmsdbs@vocmsvm06:tests$ python generatebigjson.py 20 30 400
generate dictionary:  0.714949131012
json encoding:  5.11010408401
cjson encoding:  0.642325878143
json objects are the same: True
json decoding:  12.8875789642
cjson decoding:  0.799870014191


cmsdbs@vocmsvm06:tests$ python generatebigjson.py 20 200 300
generate dictionary:  7.6853530407
json encoding:  25.7249748707
cjson encoding:  3.26199102402
json objects are the same: True
json decoding:  74.9634130001
cjson decoding:  13.6383581161


python generatebigjson.py 20 300 400
generate dictionary:  37.066849947
json encoding fails:
Traceback (most recent call last):
   File "generatebigjson.py", line 26, in <module>
     jsonobject1=json.dumps(d)
   File "/home/cmsdbs/dbs3/External/python/lib/python2.6/json/
__init__.py", line 230, in dumps
     return _default_encoder.encode(obj)
   File "/home/cmsdbs/dbs3/External/python/lib/python2.6/json/
encoder.py", line 368, in encode
     return ''.join(chunks)
MemoryError

cjson encoding:  6.63086295128
Now it's stuck on json decoding and I don't want to wait on it anymore.
```

I don't have this benchmarking script anymore (I lost couple of laptops since that year), but here is a [code](https://github.com/dmwm/DAS/blob/master/src/python/DAS/utils/jsonwrapper/__init__.py) I used in DAS python implementation which relied on 3 json's implementations, cjson, yajl and standard library json. If I recall correctly it was part of DMWM codebase at some point and eventually was removed from there.

As you can see from benchmarking numbers the cjson not only outperforms the standard library one by large margin but it also had no issue with memory footprint which was another issue we explored at that time. This issue was about large memory footprint when loading large json files. Python is really bad at this due to its memory management. Please note, that whatever you do I suggest that you perform a similar benchmarks like encoding and decoding large json files which is usually the case in DMWM codebase which quite often use dynamic keys which contributes to large memory and CPU footprints.

@yuyiguo was on that list which was exchange between me, Anzar, Simon, Dave and Aleko. So I think she will remember and find out appropriate info.

**Comment by klannon** [comment]
@vkuznet, just to make sure there is no confusion, I don't think anyone is proposing to fall back to the standard python json.  I'm just saying that there are more modern options (that probably didn't exist in 2009) that we should consider, like `orjson`, that look like they have [comparable performance](https://github.com/ijl/orjson#performance)  to cjson based on the numbers that you posted.   Hopefully we can start there.

**Comment by vkuznet** [comment]
My suggestion is to do the following (In that order):
- pick up large Phedex JSON output (as it has nested data structures)
- write benchmark for you library of choice
- run benchmark against Phedex JSON using cjson and your library of choice
- if results are comparable
  - write wrapper class similar to code I provided in DAS to avoid re-writing DBS pieces
  - write spec for deployment procedure
  - add switch to DBS to use wrapper class
  - make new release of DBS which can work with both cjson and your library of choice
  - run test with python2 and python3
If everything goes smoothly prepare DBS for switch to new library. I think you also aware that  such switch can only be done when DBS code itself will be ported to python3. So, it's going to be a lot of work. And, as I pointed out before at different occasions may be you should first evaluate if this effort is worth it since rewriting DBS in Go will eliminate all of these problems (and others), and will provide much more efficient code base which will have better lifetime in a long run and provide more benefits for time required for these efforts. It is your choice at the end.

**Comment by amaltaro** [comment]
Thanks for providing these useful suggestions, Valentin and Kevin. The check list above is close to what I have in mind.
Giving that the main goal with this issue is to be able to port *WMCore* do python3, I'd suggest to start the research/profiling of what the best library is in python3. Of course we also have to consider the results for a python2 stack, and if needed keep using a different json while our code is still running in PY2.

**Comment by mapellidario** [comment]
The investigation continued, a draft of the report is available at [https://cms-dmwm-test.docs.cern.ch/py2py3/json/](https://cms-dmwm-test.docs.cern.ch/py2py3/json/)

**Comment by amaltaro** [comment]
As discussed this morning with Dario and Todor, we do not see `cjson` dependency and/or imports within the WMCore codebase, hence, we can close this and deal with `cjson` only in DBS (which will likely keep using `cjson` in python2, and use standard library json in python3). 

Closing this as there is nothing else to be done in WMCore.
